doc/developer-notes.md

```diff
@@ -640,6 +640,29 @@ Strings and formatting
 
   - *Rationale*: Bitcoin Core uses tinyformat, which is type safe. Leave them out to avoid confusion.
 
+- Use `.c_str()` sparingly. Its only valid use is to pass C++ strings to C functions that take NULL-terminated
+  strings.
+
+  - Do not use it when passing a sized array (so along with `.size()`). Use `.data()` instead to get a pointer
+    to the raw data.
+
+    - *Rationale*: This can be dangerous, as the part of the data after the first NULL character is not guaranteed
+    to be initialized at all. Also, `.data()` communicates the intent better.
```

Comment:
Is this true after C++11? According to https://en.cppreference.com/w/cpp/string/basic_string/c_str

>  c_str() and data() perform the same function. | (since C++11)

Reply:
Ok, I didn't know that, I'll remove that from the rationale. But I'd like to keep the suggestion because I still think it's bad form to use `c_str` in these cases.